### PR TITLE
Fix handling of designator None types/values

### PR DESCRIPTION
**Pull request by alberth**
Should address #873 but not sure. @LoyVanBeek ?

### CODE REVIEW

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 1]
```diff
@@ -27,25 +27,41 @@ def __init__(self, initial_value=None, resolve_type=None, name=None):
...
+                    raise TypeError("The resolve_type may not be [None]")
+            self._resolve_type = resolve_type
+        else:
+            # No resolve type provided, use the inital value to derive a type.
+            if self.__initial_value is None:
+                raise TypeError("resolve_type could not be inferred from None.")
+
             if isinstance(self.__initial_value, list):
-                if self.__initial_value:
+                if self.__initial_value: # Not an empty list.
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]:
```suggestion
                if len(self.__initial_value) > 0:
```

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 2]
```diff
@@ -55,20 +71,32 @@ def resolve(self):
...
-            result_type = type(result[0])
+        if result is None: # No data available for checking the type.
+            return result
+
+        if isinstance(self._resolve_type, list):
+            # Not None, list expected instead.
+            if not isinstance(result, list):
+                self.fail_with_type_error(result_type, resolve_type, result)
+
+            if not result: # Empty list, cannot check element type.
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]:
```suggestion
            if len(result) == 0: # Empty list, cannot check element type.
```

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 3]
```diff
@@ -55,20 +71,32 @@ def resolve(self):
...
 
-        if result is not None and not issubclass(result_type, resolve_type):
-            raise TypeError("{} resolved to a '{}' instead of expected '{}'. "
-                            "Originals: result type: {}, resolve type: {}".format(self, result_type, resolve_type,
-                                                                                  type(result), self.resolve_type))
         return result
 
+    def fail_with_type_error(self, result_type, resolve_type, result):
+            msg = "{} resolved to a '{}' instead of expected '{}'. Originals: result type: {}, resolve type: {}"
+            raise TypeError(msg.format(self, result_type, resolve_type, type(result), self.resolve_type))
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]: resolve_type = self.resolve_type = property(_get_resolve_type) = return self._resolve_type

I think resolve type is now printed two times, both time provided via a completely different route. This should be simplified.

Both calls to `fail_with_type_error` have the same arguments, so t

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 4]
```diff
@@ -316,6 +338,55 @@ def test_list(self):
         self.assertEqual(v.resolve_type, [str])
         self.assertListEqual(v.resolve(), ['a', 'b', 'c'])
 
+    def test_none_value(self):
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]: These test are moved to the test folder in `rwc2019_robot_smach_states`, this PR should hold till that PR is merged and then these changes should be moved to the test_folder as well.

File: robot_smach_states/src/robot_smach_states/util/designators/core.py (hunk 3 above)
Comment (alberth) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Some text seems to be missing at the end?

I moved some `[0]` index operations it seems, which makes it print duplicate information more often than it did in the past. I'll look into it.

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 5]
```diff
@@ -26,25 +26,41 @@ def __init__(self, initial_value=None, resolve_type=None, name=None):
...
         super(Designator, self).__init__()
 
         self._name = name
-
         self.__initial_value = initial_value
-        if not resolve_type:
+        if resolve_type:
+            if isinstance(resolve_type, list):
+                if len(resolve_type) != 1:
+                    msg = "The resolve_type must be a sinlge-item list, found {}."
```
Comment (LoyVanBeek) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]: typo

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 6]
```diff
@@ -26,25 +26,41 @@ def __init__(self, initial_value=None, resolve_type=None, name=None):
...
         """
         super(Designator, self).__init__()
 
         self._name = name
-
         self.__initial_value = initial_value
-        if not resolve_type:
+        if resolve_type:
+            if isinstance(resolve_type, list):
+                if len(resolve_type) != 1:
```
Comment (LoyVanBeek) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]: I think there are cases where a list of types is used as 'either one of these types' but if so (and I'm not sure), then _that_ use-case is just wrong.

File: robot_smach_states/src/robot_smach_states/util/designators/core.py [hunk 7]
```diff
@@ -130,17 +158,11 @@ def __init__(self, initial_value=None, resolve_type=None, name=None):
...
         super(VariableDesignator, self).__init__(initial_value, resolve_type, name=name)
-
         self._current = initial_value
-
         self.writeable = VariableWriter(self)
 
     def _set_current_protected(self, value):
         resolve_type = self.resolve_type
-        try:
-            resolve_type = tuple(self.resolve_type)
```
Comment (LoyVanBeek) [inline robot_smach_states/src/robot_smach_states/util/designators/core.py]: WTF was this?

File: robot_smach_states/test/util/designators/test_core.py [hunk 8]
```diff
@@ -46,6 +46,55 @@ def test_list(self):
...
+            def __init__(self):
+                super(MyVariableDesignator, self).__init__(resolve_type=[object])
+            def _resolve(self):
+                return None
+
+        v = MyVariableDesignator()
+        self.assertEquals(v.resolve(), None)
+
+    def test_empty_list_value(self):
+        # Resolve a [object] to None value
```
Comment (LoyVanBeek) [inline robot_smach_states/test/util/designators/test_core.py]: Docstring doesn't match content

File: robot_smach_states/test/util/designators/test_core.py [hunk 9]
```diff
@@ -46,6 +46,55 @@ def test_list(self):
         self.assertEqual(v.resolve_type, [str])
         self.assertListEqual(v.resolve(), ['a', 'b', 'c'])
 
+    def test_none_value(self):
```
Comment (LoyVanBeek) [inline robot_smach_states/test/util/designators/test_core.py]: :heart_eyes: love all these tests!

File: robot_smach_states/src/robot_smach_states/util/designators/core.py (hunk 7 above)
Comment (alberth) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Likely caused by confusion about the semantics of "[<type>]" vs "<type>" in `resolve_type`, although I didn't check history.

Storing a non-type value in a variable that should contain a type is quite confusing, the proper solution would be to make 2 classes, one for expressing a value for a singular value, and one for expressing a value for a list of items.

File: robot_smach_states/src/robot_smach_states/util/designators/core.py (hunk 6 above)
Comment (alberth) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Clarified semantics of resolve_type somewhat

Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: The current implementation allows the following: `[[float], [int], float, int]`. So multiple resolve types are allowed. If one of the resolve types is a list, that list should contain all of the same types, so the `len([float]) == 1`.

Comment (alberth) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: In my changes, a single type means a value of that type, and list with a single type means a list of values of that type.
This is not really nice, but Python lacks the concept "list of <type>" in its type system.

As Loy says and I agree, there should not be a set of allowed types. Inheritance is supported though as it uses `isinstance`. That handles `int` vs `float` already. Mixing single values and lists of values is non-Pythonic, you should know what you get without code to figure that out.

Comment (LoyVanBeek) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Another option, if we want to allow multiple types, is to pass that asa `set()`, which is also a builtin type. 
But I do prefer to just allow 1 type to be specified. 

So, this change if fine AFAIK.

Comment (MatthijsBurgh) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Right now a mix of lists and single items is allowed. Also different single items/items in the list are allowed. So Before changing this API, please check the entire `tue_robocup` for compatibility.

File: robot_smach_states/test/util/designators/test_core.py [hunk 10]
```diff
@@ -46,6 +46,55 @@ def test_list(self):
...
+
+    def test_none_type(self):
+        with self.assertRaises(TypeError):
+            v = VariableDesignator(resolve_type=[None])
+
+    def test_multi_element_type(self):
+        with self.assertRaises(TypeError):
+            v = VariableDesignator(resolve_type=[str, bool])
+
+    def test_none_type(self):
```
Comment (LoyVanBeek) [inline robot_smach_states/test/util/designators/test_core.py]: This test name is used 2x, also on line 87

File: robot_smach_states/src/robot_smach_states/util/designators/core.py (hunk 7 above)
Comment (LoyVanBeek) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: Yes, I've been pondering to split the VariableDesignator into one for collections and one for singular values. But That's for later I'd say

File: robot_smach_states/src/robot_smach_states/util/designators/core.py (hunk 6 above)
Comment (alberth) [reply robot_smach_states/src/robot_smach_states/util/designators/core.py]: I checked the designators in tue_robocup, and found no cases of having multiple types in the constructor call. If my check was incomplete, at least it is not a commonly used feature.

**Pull request merged**